Zend/zend_property_hooks.c

```diff
@@ -36,42 +36,68 @@ typedef struct {
 static zend_result zho_it_valid(zend_object_iterator *iter);
 static void zho_it_move_forward(zend_object_iterator *iter);
 
-// FIXME: This should probably be stored on zend_class_entry somewhere (e.g. through num_virtual_props).
 static uint32_t zho_num_backed_props(zend_object *zobj)
 {
-	zend_property_info *prop_info;
-	int backed_property_count = 0;
-	ZEND_HASH_MAP_FOREACH_PTR(&zobj->ce->properties_info, prop_info) {
-		if (!(prop_info->flags & (ZEND_ACC_STATIC|ZEND_ACC_VIRTUAL))) {
-			backed_property_count++;
-		}
-	} ZEND_HASH_FOREACH_END();
-	return backed_property_count;
+	return zobj->ce->default_properties_count;
 }
 
-static zend_array *zho_build_properties_ex(zend_object *zobj, bool check_access, bool include_dynamic_props)
+static zend_array *zho_build_properties_ex(zend_object *zobj, bool check_access, bool force_ptr, bool include_dynamic_props)
 {
 	zend_class_entry *ce = zobj->ce;
 	zend_array *properties = zend_new_array(ce->default_properties_count);
 	zend_hash_real_init_mixed(properties);
 
-	zend_property_info *prop_info;
-	ZEND_HASH_MAP_FOREACH_PTR(&ce->properties_info, prop_info) {
-		if (prop_info->flags & ZEND_ACC_STATIC) {
-			continue;
-		}
-		if (check_access && zend_check_property_access(zobj, prop_info->name, false) == FAILURE) {
-			continue;
-		}
-		if (prop_info->hooks) {
-			_zend_hash_append_ptr(properties, prop_info->name, prop_info);
-		} else {
-			if (UNEXPECTED(Z_TYPE_P(OBJ_PROP(zobj, prop_info->offset)) == IS_UNDEF)) {
-				HT_FLAGS(properties) |= HASH_FLAG_HAS_EMPTY_IND;
+	/* Build list of parents */
+	int32_t parent_count = 0;
+	for (zend_class_entry *pce = ce; pce; pce = pce->parent) {
+		parent_count++;
+	}
+	zend_class_entry **parents = emalloc(sizeof(zend_class_entry*) * parent_count);
+	int32_t i = 0;
+	for (zend_class_entry *pce = ce; pce; pce = pce->parent) {
+		parents[i++] = pce;
+	}
+
+	/* Iterate parents top to bottom */
+	i--;
+	for (; i >= 0; i--) {
+		zend_class_entry *pce = parents[i];
+
+		zend_property_info *prop_info;
+		ZEND_HASH_MAP_FOREACH_PTR(&pce->properties_info, prop_info) {
+			if (prop_info->flags & ZEND_ACC_STATIC) {
+				continue;
 			}
-			_zend_hash_append_ind(properties, prop_info->name, OBJ_PROP(zobj, prop_info->offset));
-		}
-	} ZEND_HASH_FOREACH_END();
+			zend_string *property_name = prop_info->name;
+			/* When promoting properties from protected to public, use the unmangled name to preserve order. */
+			if (prop_info->flags & ZEND_ACC_PROTECTED) {
```

Comment:
This causes `zend_check_property_access()` to fail when the property is protected and not promoted to public.